src/app/app-platform/ContentAppPlatform.cpp

```diff
@@ -100,37 +100,67 @@ EndpointId ContentAppPlatform::AddContentApp(ContentApp * app, EmberAfEndpointTy
         }
         index++;
     }
 
+    if (desiredEndpointId >= emberAfFixedEndpointCount() &&
+        emberAfGetDynamicIndexFromEndpoint(desiredEndpointId) != kEmberInvalidEndpointIndex)
+    {
+        // an endpoint already exists with the desiredEndpointId. return error.
+        ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: desired endpointID already in use!");
+        return kNoCurrentEndpointId;
+    }
+
     index = 0;
     while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
     {
         if (nullptr == mContentApps[index])
         {
             mContentApps[index] = app;
             EmberAfStatus ret;
-            while (1)
+
+            if (desiredEndpointId >= emberAfFixedEndpointCount())
             {
-                ret = emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, dataVersionStorage, deviceTypeList);
+                ret = emberAfSetDynamicEndpoint(index, desiredEndpointId, ep, dataVersionStorage, deviceTypeList);
                 if (ret == EMBER_ZCL_STATUS_SUCCESS)
                 {
                     ChipLogProgress(DeviceLayer, "Added ContentApp %s to dynamic endpoint %d (index=%d)", vendorApp.applicationId,
-                                    mCurrentEndpointId, index);
-                    app->SetEndpointId(mCurrentEndpointId);
+                                    desiredEndpointId, index);
+                    app->SetEndpointId(desiredEndpointId);
                     return app->GetEndpointId();
                 }
-                else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS)
+                else
                 {
                     ChipLogError(DeviceLayer, "Adding ContentApp error=%d", ret);
                     return kNoCurrentEndpointId;
                 }
-                // Handle wrap condition
-                if (++mCurrentEndpointId < mFirstDynamicEndpointId)
+            }
+            else
+            {
+                while (1)
                 {
-                    mCurrentEndpointId = mFirstDynamicEndpointId;
+                    ret = emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, dataVersionStorage, deviceTypeList);
+                    if (ret == EMBER_ZCL_STATUS_SUCCESS)
+                    {
+                        ChipLogProgress(DeviceLayer, "Added ContentApp %s to dynamic endpoint %d (index=%d)",
+                                        vendorApp.applicationId, mCurrentEndpointId, index);
+                        app->SetEndpointId(mCurrentEndpointId);
+                        return app->GetEndpointId();
+                    }
+                    else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS)
+                    {
+                        ChipLogError(DeviceLayer, "Adding ContentApp error=%d", ret);
+                        return kNoCurrentEndpointId;
+                    }
+                    // Handle wrap condition
+                    if (++mCurrentEndpointId < mFirstDynamicEndpointId)
```

Comment:
Is there a guarantee that we do not always return `EMBER_ZCL_STATUS_DUPLICATE_EXISTS`? What guarantees no infinite loop here? please explain in a comment.

Reply:
This code was not changed in this PR. Under the assumption that only the content app platform manages the endpoints there will be no infinite loop since as long as there are app slots there will be endpoints available. But the code will break if the assumption does not hold. which is that other things can also add/remove dynamic endpoints. updated code.